hail/src/main/scala/is/hail/expr/types/virtual/TTuple.scala

```diff
@@ -22,23 +22,19 @@ object TTuple {
 }
 
 final case class TTuple(_types: IndexedSeq[Type], override val required: Boolean = false) extends TBaseStruct {
-  lazy val physicalType: PTuple = PTuple(types.map(_.physicalType), required)
+  lazy val physicalType: PTuple = PTuple(_types.map(_.physicalType), required)
 
-  val types = _types.toArray
-  val fieldRequired: Array[Boolean] = types.map(_.required)
+  lazy val types = _types.toArray
 
-  val fields: IndexedSeq[Field] = types.zipWithIndex.map { case (t, i) => Field(s"$i", t, i) }
+  lazy val fields: IndexedSeq[Field] = types.zipWithIndex.map { case (t, i) => Field(s"$i", t, i) }
 
-  val ordering: ExtendedOrdering = TBaseStruct.getOrdering(types)
+  lazy val ordering: ExtendedOrdering = TBaseStruct.getOrdering(types)
 
-  val size: Int = types.length
+  def size: Int = types.length
```

Comment:
If types is lazy, can its length/size be lazy as well?

Reply:
it can, but this will be just as fast and will slightly reduce the in-memory footprint of the object.